packages/main/src/Tree.js

```diff
@@ -0,0 +1,378 @@
+			},
+		},
+
+		/**
+		 * Fired when a tree item is activated.
+		 *
+		 * @event
+		 * @param {HTMLElement} item the clicked item.
+		 * @public
+		 */
+		itemClick: {
+			detail: {
+				item: { type: HTMLElement },
+			},
+		},
+
+		/**
+		 * Fired when the Delete button of any tree item is pressed.
+		 * <br><br>
+		 * <b>Note:</b> A Delete button is displayed on each item,
+		 * when the <code>ui5-tree</code> <code>mode</code> property is set to <code>Delete</code>.
+		 * @event
+		 * @param {HTMLElement} item the deleted item.
+		 * @public
+		 */
+		itemDelete: {
+			detail: {
+				item: { type: HTMLElement },
+			},
+		},
+
+		/**
+		 * Fired when selection is changed by user interaction
+		 * in <code>SingleSelect</code>, <code>SingleSelectBegin</code>, <code>SingleSelectEnd</code> and <code>MultiSelect</code> modes.
+		 *
+		 * @event
+		 * @param {Array} selectedItems An array of the selected items.
+		 * @param {Array} previouslySelectedItems An array of the previously selected items.
+		 * @public
+		 */
+		selectionChange: {
+			detail: {
+				selectedItems: { type: Array },
+				previouslySelectedItems: { type: Array },
+			},
+		},
+	},
+};
+
+/**
+ * @class
+ *
+ * <h3 class="comment-api-title">Overview</h3>
+ * The <code>ui5-tree</code> component provides a tree structure for displaying data in a hierarchy.
+ *
+ * <h3>Keyboard Handling</h3>
+ * <code>ui5-tree</code> provides advanced keyboard handling. You can use the up/down arrow keys to navigate to the previous/next item in the list,
+ * representing the tree, regardless of nesting, but also the left/right arrow keys to drill down and go up the tree. If you press the right arrow
+ * on a tree node, it will expand, if not expanded. If you press the right arrow key once more, the first child of this node will be selected.
+ * If you press the left arrow on a tree node, it will collapse, if expanded. If you press the left arrow key once more, the parent node of this
+ * tree node will be selected.
+ *
+ * <h3>ES6 Module Import</h3>
+ * <code>import @ui5/webcomponents/dist/Tree.js";</code>
+ * <br>
+ * <code>import @ui5/webcomponents/dist/TreeItem.js";</code> (for its respective item element)
+ *
+ * @constructor
+ * @author SAP SE
+ * @alias sap.ui.webcomponents.main.Tree
+ * @extends UI5Element
+ * @tagname ui5-tree
+ * @appenddocs TreeItem
+ * @public
+ * @since 1.0.0-rc.8
+ */
+class Tree extends UI5Element {
+	static get metadata() {
+		return metadata;
+	}
+
+	static get render() {
+		return litRender;
+	}
+
+	static get styles() {
+		return TreeCss;
+	}
+
+	static get template() {
+		return TreeTemplate;
+	}
+
+	static async onDefine() {
+		await Promise.all([
+			BusyIndicator.define(),
+			List.define(),
+			TreeListItem.define(),
+			TreeItem.define(),
+		]);
+	}
+
+	onBeforeRendering() {
+		this._listItems = [];
+		buildTree(this, 1, this._listItems);
+	}
+
+	onEnterDOM() {
+		this._observer = new MutationObserver(this.onTreeStructureChange.bind(this));
```

Comment:
Perhaps we can check if the _observer is available and not create multiple observers if the tree is added and removed from the DOM

Reply:
Done, moved to constructor